File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/region/ComponentPath.ts
module api.content.page.region {

    export class ComponentPath implements api.Equitable {

        private static DIVIDER = "/";

        private regionAndComponentList: ComponentPathRegionAndComponent[];

        private refString: string;

        constructor(regionAndComponentList: ComponentPathRegionAndComponent[]) {

            this.regionAndComponentList = regionAndComponentList;

            this.refString = "";
            this.regionAndComponentList.forEach((regionAndComponent: ComponentPathRegionAndComponent, index: number) => {
                this.refString += regionAndComponent.toString();
                if (index < this.regionAndComponentList.length - 1) {
                    this.refString += ComponentPath.DIVIDER;
                }
            });
        }

        numberOfLevels(): number {
            return this.regionAndComponentList.length;
        }

        getFirstLevel(): ComponentPathRegionAndComponent {
            return this.regionAndComponentList[0];
        }

        getLastLevel(): ComponentPathRegionAndComponent {
            return this.regionAndComponentList[this.regionAndComponentList.length - 1];
        }

        getLevels(): ComponentPathRegionAndComponent [] {
            return this.regionAndComponentList;
        }

        getComponentIndex(): number {
            return this.getLastLevel().getComponentIndex();
        }

        getRegionPath(): RegionPath {

            var regionPathAsString = "";
            this.regionAndComponentList.forEach((regionAndComponent: ComponentPathRegionAndComponent, index: number) => {

                if (index == this.regionAndComponentList.length - 1) {
                    regionPathAsString += regionAndComponent.getRegionName();
                }
                else {
                    regionPathAsString += regionAndComponent.toString();
                    regionPathAsString += "/";
                }

            });

            return RegionPath.fromString(regionPathAsString);
        }

        public removeFirstLevel(): ComponentPath {
            if (this.numberOfLevels() <= 1) {
                return null;
            }

            var newRegionAndComponentList: ComponentPathRegionAndComponent[] = [];
            for (var i = 1; i < this.regionAndComponentList.length; i++) {
                newRegionAndComponentList.push(this.regionAndComponentList[i]);
            }
            return new ComponentPath(newRegionAndComponentList);
        }

        public toString(): string {
            return this.refString;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ComponentPath)) {
                return false;
            }

            var other = <ComponentPath>o;

            if (!api.ObjectHelper.stringEquals(this.refString, other.refString)) {
                return false;
            }

            return true;
        }

        public static fromString(str: string): ComponentPath {

            if (!str) {
                return null;
            }

            var elements: string[] = api.util.StringHelper.removeEmptyStrings(str.split(ComponentPath.DIVIDER));

            var regionAndComponentList: ComponentPathRegionAndComponent[] = [];
            for (var i = 0; i < elements.length - 1; i += 2) {
                var regionName = elements[i];
                var componentIndexAsString = elements[i + 1];
                var regionAndComponent = new ComponentPathRegionAndComponent(regionName, parseInt(componentIndexAsString));
                regionAndComponentList.push(regionAndComponent);
            }

            return new ComponentPath(regionAndComponentList);
        }

        public static fromRegionPathAndComponentIndex(regionPath: RegionPath, componentIndex: number): ComponentPath {
            api.util.assertNotNull(regionPath, "regionPath cannot be null");
            api.util.assert(componentIndex >= 0, "componentIndex must be zero or more");

            var regionAndComponentList: ComponentPathRegionAndComponent[] = [];
            if (regionPath.getParentComponentPath()) {
                regionPath.getParentComponentPath().regionAndComponentList.forEach((regionAndComponent: ComponentPathRegionAndComponent)=> {
                    regionAndComponentList.push(regionAndComponent);
                });
            }
            regionAndComponentList.push(new ComponentPathRegionAndComponent(regionPath.getRegionName(), componentIndex));
            return new ComponentPath(regionAndComponentList);
        }
    }

    export class ComponentPathRegionAndComponent {

        private static DIVIDER = "/";

        private regionName: string;

        private componentIndex: number;

        private refString: string;

        constructor(regionName: string, componentIndex: number) {
            this.regionName = regionName;
            this.componentIndex = componentIndex;
            this.refString = regionName + ComponentPathRegionAndComponent.DIVIDER + this.componentIndex;
        }

        getRegionName(): string {
            return this.regionName;
        }

        getComponentIndex(): number {
            return this.componentIndex;
        }

        toString(): string {
            return this.refString;
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/region/ComponentPath.java
package com.enonic.xp.region;

import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;

import com.enonic.xp.annotation.PublicApi;

@PublicApi
public final class ComponentPath
    implements Iterable<ComponentPath.RegionAndComponent>
{
    public static final String DIVIDER = "/";

    private final ImmutableList<RegionAndComponent> regionAndComponentList;

    public ComponentPath( final ImmutableList<RegionAndComponent> regionAndComponentList )
    {
        this.regionAndComponentList = regionAndComponentList;
    }

    public static ComponentPath from( final RegionPath parentPath, final int componentIndex )
    {
        final ImmutableList.Builder<RegionAndComponent> builder = new ImmutableList.Builder<>();
        if ( parentPath.getParentComponentPath() != null )
        {
            builder.addAll( parentPath.getParentComponentPath().regionAndComponentList );
        }
        builder.add( RegionAndComponent.from( parentPath.getRegionName(), componentIndex ) );
        return new ComponentPath( builder.build() );
    }

    public static ComponentPath from( final String str )
    {
        final List<String> valueList = Splitter.on( DIVIDER ).omitEmptyStrings().splitToList( str );

        Preconditions.checkArgument( valueList.size() % 2 == 0, "Expected even number of path elements: " + str );

        final ImmutableList.Builder<RegionAndComponent> builder = new ImmutableList.Builder<>();
        for ( int i = 0; i < valueList.size() - 1; i += 2 )
        {
            final RegionName regionName = RegionName.fromString( valueList.get( i ) );
            final ComponentIndex componentIndex = ComponentIndex.fromString( valueList.get( i + 1 ) );
            builder.add( new RegionAndComponent( regionName, componentIndex ) );

        }
        return new ComponentPath( builder.build() );
    }

    public int getComponentIndex()
    {
        return getLastLevel().getComponentIndex();
    }

    public boolean isEmpty()
    {
        return this.regionAndComponentList.isEmpty();
    }

    int numberOfLevels()
    {
        return this.regionAndComponentList.size();
    }

    RegionAndComponent getFirstLevel()
    {
        return this.regionAndComponentList.get( 0 );
    }

    RegionAndComponent getLastLevel()
    {
        return this.regionAndComponentList.get( regionAndComponentList.size() - 1 );
    }

    ComponentPath removeFirstLevel()
    {
        if ( this.numberOfLevels() <= 1 )
        {
            return null;
        }

        final ImmutableList.Builder<RegionAndComponent> builder = new ImmutableList.Builder<>();
        for ( int i = 1; i < this.regionAndComponentList.size(); i++ )
        {
            builder.add( this.regionAndComponentList.get( i ) );
        }
        return new ComponentPath( builder.build() );
    }

    @Override
    public Iterator<RegionAndComponent> iterator()
    {
        return regionAndComponentList.iterator();
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof ComponentPath ) )
        {
            return false;
        }
        final ComponentPath that = (ComponentPath) o;
        return regionAndComponentList.equals( that.regionAndComponentList );
    }

    @Override
    public int hashCode()
    {
        return regionAndComponentList.hashCode();
    }

    @Override
    public String toString()
    {
        return this.regionAndComponentList.stream().map( Objects::toString ).collect( Collectors.joining( DIVIDER, DIVIDER, "" ) );
    }

    public static class RegionAndComponent
    {
        private static final String DIVIDER = "/";

        private final RegionName regionName;

        private final ComponentIndex componentIndex;

        public RegionAndComponent( final RegionName regionName, final ComponentIndex componentIndex )
        {
            this.regionName = regionName;
            this.componentIndex = componentIndex;
        }

        public static RegionAndComponent from( final String regionName, final int componentIndex )
        {
            return new RegionAndComponent( new RegionName( regionName ), new ComponentIndex( componentIndex ) );
        }

        public static RegionAndComponent from( final String str )
        {
            final Iterable<String> values = Splitter.on( DIVIDER ).omitEmptyStrings().split( str );
            final Iterator<String> iterator = values.iterator();
            final RegionName regionName = RegionName.fromString( iterator.next() );
            final ComponentIndex componentIndex = ComponentIndex.fromString( iterator.next() );
            return new RegionAndComponent( regionName, componentIndex );
        }

        public String getRegionName()
        {
            return regionName.name;
        }

        public int getComponentIndex()
        {
            return componentIndex.index;
        }

        @Override
        public String toString()
        {
            return this.regionName + DIVIDER + this.componentIndex;
        }
    }

    public static class RegionName
    {
        private final String name;

        public RegionName( final String name )
        {
            this.name = name;
        }

        public static RegionName fromString( final String str )
        {
            return new RegionName( str );
        }

        @Override
        public String toString()
        {
            return this.name;
        }
    }

    public static class ComponentIndex
    {
        private final int index;

        public ComponentIndex( final int index )
        {
            this.index = index;
        }

        public static ComponentIndex fromString( final String str )
        {
            return new ComponentIndex( Integer.parseInt( str ) );
        }

        @Override
        public String toString()
        {
            return Integer.toString( this.index );
        }
    }
}
